### DISCUSSION TITLE
Retrieve queries based on metadata in a Qdrant database

### DISCUSSION THREAD

**Discussion by Mohana-Murugan** [qdrant/qdrant#5594] [Q&A]
Hi,

I am working on a RAG-based LLM agent. Here, I am storing multiple user details and document details in a single collection, organized based on user_id and agent_id. I am writing a search approach to retrieve information from the database based on these IDs. However, I am facing an issue where the results return information even if the specified information is not in the document. Additionally, how can I isolate data when the same files are stored for different users?

''' query = "Bullying"
query_vector = get_embedding(query)

# Search with a filter to match only documents containing "Qdrant"
results = client.search(
    collection_name='test',
    query_vector=query_vector,
    limit=1,
    query_filter={
        "must": [
            {"key": "user_id", "match": {"value": "K5617"}},
            {"key": "agent_id", "match": {"value": "5307"}},
        ]
    }
)

# Check if results contain any data
if results:
    for result in results:
        print(f"Document ID: {result.id}, Score: {result.score}")
        print(f"Text: {result.payload['text']}\n")
else:
    print("Information not available")'''

**Comment by timvisee** [comment]
> However, I am facing an issue where the results return information even if the specified information is not in the document.

I'm not exactly sure what your approach is like, because your example does not make that very clear.

If you're using payload values to isolate searches, that should already be sufficient. If it is problematic somehow I expect there to be a bug in the implementation.

If you're embedding your query and are searching using vector similarity, results can be approximate. For example, if there are no close matches to your search query, it'll simply return the closest match it can find even if it's completely unrelated material.

> Additionally, how can I isolate data when the same files are stored for different users?

You can isolate data by specifying a payload value that is unique across users. Please make sure you configure a payload index for this, as it'll help to keep searches fast and to keep good search accuracy.

**Reply by Mohana-Murugan** [reply]
Hi, thank you for your response. 

My requirement is that I want to store different user documents based on their ID and their associated agent ID. Similarly, many users will be present, each with their own agent ID. I want to isolate the information based on the user's ID and agent ID. Users should pass their ID and agent ID in the query to retrieve their specific information. If the ID matches, it should return the answer; otherwise, it should return 'Information not available.'

The problem I am facing with Qdrant is that if the ID does not match based on the filter, it gives irrelevant information instead of returning empty values.

"""
def store_embeddings_in_qdrant(text_chunks, user_id, company_id, agent_id):
    """
    Generates embeddings and stores them in Qdrant with specific metadata for access control.

    Args:
        text_chunks (list): List of text chunks to embed and store.
        user_id (str): User ID associated with the embeddings.
        company_id (str): Company ID associated with the embeddings.
        agent_id (str): Agent ID associated with the embeddings.
    """
    # Initialize embedding model
    embeddings_model = HuggingFaceEmbeddings(model_name='BAAI/bge-small-en-v1.5')
    embeddings = embeddings_model.embed_documents([chunk.page_content for chunk in text_chunks])

    # Prepare points for Qdrant
    points = []
    for idx, chunk in enumerate(text_chunks):
        point_id = str(uuid.uuid4())  # Generate unique ID for each point
        points.append(PointStruct(
            id=point_id,
            vector=embeddings[idx],
            payload={
                "text": chunk.page_content,
                "source": chunk.metadata["source"],
                "user_id": user_id,
                "company_id": company_id,
                "agent_id": agent_id
            }
        ))

    collection_name = "custom_agent"

    client = QdrantClient(url=qdrant_url, api_key=qdrant_api_key)

    # Create collection if it doesn't exist
    try:
        client.get_collection(collection_name=collection_name)
    except Exception:
        # If collection does not exist, create a new one
        client.create_collection(
            collection_name=collection_name,
            vectors_config=models.VectorParams(size=len(points[0].vector), distance=models.Distance.COSINE),
        )

    client.upsert(collection_name=collection_name, points=points)
    print(f"Successfully stored {len(points)} points in Qdrant collection '{collection_name}'.")

store_embeddings_in_qdrant(chunks, user_id, company_id, agent_id)

"""

I have stored this at Qdrant but I am not able to retrieve information based filter condition. 

without a filter, it provides the correct answer 

""" # Initialize embeddings model
embeddings_model = HuggingFaceEmbeddings(model_name='BAAI/bge-small-en-v1.5')

# Initialize vector store from existing collection
vectorstore = Qdrant(
    client=qdrant_client,
    collection_name="custom_agent",
    embeddings=embeddings_model,
)

# Initialize retriever
retriever = vectorstore.as_retriever(
    search_type="similarity",
    search_kwargs={
        "k": 5  # Top-k results
    }
)

# Initialize ChatGroq LLM
llm = ChatGroq(
    groq_api_key=groq_api_key, 
    model_name="llama3-8b-8192",
)

# Create a custom prompt template (optional)
prompt_template = """
You are a helpful assistant that uses the provided context to answer the user's question.
If you don't know the answer, say that you don't know.

Context: {context}
Question: {question}
Helpful Answer:
"""

prompt = PromptTemplate(
    input_variables=["context", "question"],
    template=prompt_template,
)

# Create a conversational retrieval chain
memory = ConversationBufferMemory(memory_key="chat_history", return_messages=True)
qa_chain = ConversationalRetrievalChain.from_llm(
    llm=llm,
    retriever=retriever,
    memory=memory,
    verbose=True,
    combine_docs_chain_kwargs={"prompt": prompt},
)
"""

![pic-1](https://github.com/user-attachments/assets/e51bea74-8be3-4e39-a5bb-7eccef2716d0)

with filter, it provide answers below 

""" # Initialize embeddings model
embeddings_model = HuggingFaceEmbeddings(model_name='BAAI/bge-small-en-v1.5')

# Initialize vector store from existing collection
vectorstore = Qdrant(
    client=qdrant_client,
    collection_name="custom_agent",
    embeddings=embeddings_model,
)

metadata_filter = {
    "must": [
        {"key": "user_id", "match": {"value": "K5617"}},
        {"key": "agent_id", "match": {"value": "5307"}},
        {"key": "company_id", "match": {"value": "101"}},
    ]
}

# Initialize retriever
retriever = vectorstore.as_retriever(
    search_type="similarity",
    search_kwargs={
        "k": 5,  # Top-k results
        "filter": metadata_filter  # Apply metadata filter
    },
)

# Initialize ChatGroq LLM
llm = ChatGroq(
    groq_api_key=groq_api_key, 
    model_name="llama3-8b-8192",
)

# Create a custom prompt template (optional)
prompt_template = """
You are a helpful assistant that uses the provided context to answer the user's question based on documents available in the database.
If you don't know the answer, explicitly state that you don't know.

Context: {context}
Question: {question}

Helpful Answer:

"""

prompt = PromptTemplate(
    input_variables=["context", "question"],
    template=prompt_template,
)

# Create a conversational retrieval chain
memory = ConversationBufferMemory(memory_key="chat_history", return_messages=True)
qa_chain = ConversationalRetrievalChain.from_llm(
    llm=llm,
    retriever=retriever,
    memory=memory,
    verbose=True,
    combine_docs_chain_kwargs={"prompt": prompt},
)
"""

![pic-1 (1)](https://github.com/user-attachments/assets/d3726693-597f-43c4-b609-b6370020f3f4)

**Reply by timvisee** [reply]
Thank you for all the details! I only skimmed it though, because it's quite a bit to read through.

> The problem I am facing with Qdrant is that if the ID does not match based on the filter, it gives irrelevant information instead of returning empty values.

Could it be that it finds results with low similarity, but still matching the provided ID and agent ID?

If so, that's expected. It'll always return those because Qdrant will still consider those as 'most similar' even though they might not appear similar at all according to you.

**Reply by Mohana-Murugan** [reply]
Yes, I am not able to get the right answer based on a filter.

**Reply by Mohana-Murugan** [reply]
So, we have planned to move Pinecone vector DB because of these issues.

**Reply by timvisee** [reply]
If what I describe above is what's happening - Pinecone will not fix this for you and give you exactly the same behavior.

Can you reduce the problem you're seeing to an example snippet of about 10 lines of code? Then I'd be happy to help you out with this.

For example, if we take the following snippet from your original comment, what response does Qdrant give you?

```
results = client.search(
collection_name='test',
query_vector=query_vector,
limit=1,
query_filter={
"must": [
{"key": "user_id", "match": {"value": "K5617"}},
{"key": "agent_id", "match": {"value": "5307"}},
]
}
)
```

Up until now it's still not clear to me what the exact issue is. Because I'm afraid that 'I am not able to get the right answer based on a filter' is rather vague.

**Reply by Mohana-Murugan** [reply]
My exact issues are as follows
- I have stored the RAG PDF under: user_id: K5617, Agent_id: 5307, Company ID: 101. 
- I have stored the HR Policies under: user_id: K5618, Agent_id: 5308, Company ID: 102

It works fine:
![pic-1](https://github.com/user-attachments/assets/d978cff9-40c7-47ee-9fa9-1289751c61f1)
![pic-1](https://github.com/user-attachments/assets/6f01e109-3ef8-4c37-8c76-58af710cd6bb)

When IDs do not match:
![pic-1](https://github.com/user-attachments/assets/50c9ca34-f13c-4169-87da-cd8a1fe4d65a)

It provides the following wrong answer 
![pic-1](https://github.com/user-attachments/assets/f41474ae-80c9-4794-9f17-e3a1dbbf77ab)

**Reply by Mohana-Murugan** [reply]
I am using the same embedding model, 'BAAI/bge-small-en-v1.5,' for storing and retrieving. However, I cannot identify the reason why I am getting these results.

When I pass this to the LLM, the model provides answers based on its pre-trained documents instead of retrieving information from Qdrant using a filter.

Without any filter, Qdrant works fine for us, and there is no doubt about that.  (I have attached the results of with and without LLM  and filter condition above)

If you provide a solution for this issue that would be helpful for us. Million thanks in advance..!
